### PR TITLE
feat: add temporal query parameters to MCP tools

### LINKED ISSUE #511: [FR] Add timeframe/temporal query parameters to MCP tools for trend analysis
## Problem Statement

Currently, Sourcebot's MCP tools (`search_code`, `list_repos`, `get_file_source`) only provide access to the current state of the codebase. This limitation prevents AI language models from performing meaningful temporal analysis, trend reporting, or historical insights when using the Sourcebot MCP server.

## Use Case

When LLMs attempt to analyze code evolution, track changes over time, or provide trend analysis, they are unable to do so effectively because the MCP tools lack timeframe filtering capabilities. This results in responses like "I cannot analyze trends over time with the available tools" or incomplete analysis based only on current code state.

### Real-World Examples:
- **Trend Analysis**: "Show me how the usage of async/await has evolved in our codebase over the last year"
- **Historical Search**: "Find all instances where deprecated function X was removed between March and June"
- **Evolution Tracking**: "Compare the error handling patterns used 6 months ago vs today"
- **Regression Analysis**: "Identify when specific security patterns were introduced across repositories"

## Proposed Solution

Add temporal/timeframe parameters to existing MCP tools to enable time-based querying:

### Enhanced `search_code` Tool
Add optional parameters:
```typescript
{
  since?: string;          // ISO date string (e.g., "2024-01-01")
  until?: string;          // ISO date string (e.g., "2024-12-31")
  dateRange?: string;      // Relative range (e.g., "last-3-months", "last-year")
  gitRevision?: string;    // Specific commit/tag/branch
  compareRevisions?: {     // Compare between two points in time
    from: string;
    to: string;
  };
  includeDeletedFiles?: boolean; // Include files that were deleted in timeframe
}
```

### Enhanced `list_repos` Tool
Add repository filtering by activity timeframe:
```typescript
{
  activeAfter?: string;    // Only repos with commits after this date
  activeBefore?: string;   // Only repos with commits before this date
}
```

### New `search_commits` Tool
```typescript
{
  query: string;           // Search in commit messages, authors, changes
  since?: string;
  until?: string;
  author?: string;
  repository?: string[];
}
```

## Technical Implementation Considerations

1. **Git Integration**: Leverage git log with `--since`, `--until`, `--author` flags
2. **Zoekt Enhancement**: Extend Zoekt indexing to include commit metadata and timestamps
3. **API Compatibility**: Maintain backward compatibility by making all new parameters optional
4. **Performance**: Consider caching mechanisms for frequently requested time ranges
5. **Date Formats**: Support both absolute (ISO 8601) and relative dates ("1 week ago", "last month")

## Benefits

- **Enhanced AI Analysis**: Enable LLMs to provide comprehensive temporal insights and trend analysis
- **Historical Context**: Allow developers to understand code evolution and decision history
- **Compliance & Auditing**: Support regulatory requirements for tracking code changes over time
- **Research Capabilities**: Enable academic and organizational research on codebase evolution
- **Debugging**: Help identify when bugs or issues were introduced

## Related Issues

This request complements issue #22 "FR: Ability to query for configuration in repositories over time" but focuses specifically on enhancing the MCP interface for broader temporal querying capabilities.

## Community Impact

This feature would make Sourcebot's MCP integration significantly more powerful for AI-assisted development workflows, positioning it as a leader in temporal code analysis capabilities. Many organizations need this functionality for compliance, research, and advanced AI-powered insights.

## Alternative Approaches Considered

1. **Client-side Git Integration**: Less efficient and requires additional tooling
2. **Separate Temporal API**: Would fragment the interface and require additional complexity
3. **External Time-series Database**: Overkill for this use case and adds infrastructure complexity

The proposed approach of enhancing existing MCP tools provides the most seamless integration with current workflows while enabling powerful new capabilities.

### PR DESCRIPTION
**Pull request by waynesun09**
- Added endpoint for searching commit history for a git repository.
- [mcp] Added `search_commits` tool to search a repos commit history.
- [mcp] Added `gitRevision` parameter to the `search_code` tool to allow for searching on different branches.

Fixes #511

### CODE REVIEW

**Comment by brendan-kellam** [comment]
@coderabbitai review

File: packages/web/src/features/search/schemas.ts [hunk 1]
```diff
@@ -30,6 +30,14 @@ export const searchRequestSchema = z.object({
...
     // Whether to return the whole file as part of the response.
     whole: z.boolean().optional(),
+    // The git revision to search in.
+    gitRevision: z.string().optional(),
+    // Filter by date. Only show results that were indexed after this date.
+    since: z.string().optional(),
+    // Filter by date. Only show results that were indexed before this date.
+    until: z.string().optional(),
+    // Whether to include deleted files in the search results.
+    includeDeletedFiles: z.boolean().optional(),
```
Comment (brendan-kellam) [inline packages/web/src/features/search/schemas.ts]: Seems like this is unused?

File: packages/web/src/features/search/searchApi.ts [hunk 2]
```diff
@@ -127,8 +127,70 @@ const getFileWebUrl = (template: string, branch: string, fileName: string): stri
...
 
-export const search = async ({ query, matches, contextLines, whole }: SearchRequest): Promise<SearchResponse | ServiceError> => sew(() =>
+export const search = async ({ query, matches, contextLines, whole, gitRevision, since, until, includeDeletedFiles }: SearchRequest): Promise<SearchResponse | ServiceError> => sew(() =>
     withOptionalAuthV2(async ({ org, prisma }) => {
+        // If gitRevision is provided, append it to the query as a branch filter.
+        if (gitRevision) {
+            query = `${query} branch:${gitRevision}`;
+        }
+
+        // If since or until are provided, filter repositories by indexedAt.
```
Comment (brendan-kellam) [inline packages/web/src/features/search/searchApi.ts]: I can see the value in being able to search the commit time, but what is the benefit of being able to filter search results by indexedAt?

File: packages/web/src/actions.ts [hunk 3]
```diff
@@ -460,23 +460,75 @@ export const getUserApiKeys = async (domain: string): Promise<{ name: string; cr
...
 }: {
     where?: Prisma.RepoWhereInput,
-    take?: number
+    take?: number,
+    activeAfter?: string,
+    activeBefore?: string,
 } = {}) => sew(() =>
     withOptionalAuthV2(async ({ org, prisma }) => {
+        // When filtering by activity, we need to fetch all repos first,
+        // then apply pagination after filtering
```
Comment (brendan-kellam) [inline packages/web/src/actions.ts]: nit: we aren't really doing pagination here with `take`

File: packages/web/src/features/search/gitApi.ts [hunk 4]
```diff
@@ -0,0 +1,150 @@
...
+ *
+ * **Requirements**: The repository must be cloned on the Sourcebot server disk.
+ * Sourcebot automatically clones repositories during indexing, but the cloning
+ * process might not be finished when this query is executed. If the repository
+ * is not found on the server disk, an error will be returned.
+ *
+ * @param request - Search parameters including timeframe filters
+ * @returns Array of commits or ServiceError
+ */
+export const searchCommits = async ({
```
Comment (brendan-kellam) [inline packages/web/src/features/search/gitApi.ts:26]: Since this is exposed in the api surface (in `route.ts`), it will need a auth guard (e.g., `withOptionalAuthV2`)

File: packages/web/src/features/search/schemas.ts (hunk 1 above)
Comment (waynesun09) [reply packages/web/src/features/search/schemas.ts]: You are right, the function does not work as query is not supported also, I'll remove it.

File: packages/web/src/features/search/searchApi.ts (hunk 2 above)
Comment (waynesun09) [reply packages/web/src/features/search/searchApi.ts]: My initial plan is to add temporal option in search_code, but turns out it's better implement to provide new search_commit tool specifically to get the latest code change with commit time. 

For the indexedAt option of search_code, the usage could be targeting two scenarios:
- New Repos: If you add a new repository to Sourcebot today, it will have an indexedAt timestamp of today. Searching with since="today" will immediately include code from this new repository, allowing you to answer "What code was just added to our knowledge base?"
- Synced Repos: When latest changes for an existing repo and re-indexes it, the indexedAt timestamp is updated.
Scenario: A developer pushes a critical security fix and re-indexes happened on the repo.
Result: A search with since="1 hour ago" will surface matches from that repo because it was just re-processed, effectively showing you "freshly indexed code."

File: packages/web/src/actions.ts (hunk 3 above)
Comment (waynesun09) [reply packages/web/src/actions.ts]: Ok, thanks for clarify that.
My implementation with remove the limit to load as much as possible might cause server side performance issue, so I'll update it also with combine the take at end, but batching load first and set hard maximum limit on total.

File: packages/web/src/features/search/gitApi.ts (hunk 4 above)
Comment (waynesun09) [reply packages/web/src/features/search/gitApi.ts:26]: OK

File: packages/web/src/features/search/schemas.ts (hunk 1 above)
[the remark this answers is not in the document]
Comment (waynesun09) [reply packages/web/src/features/search/schemas.ts]: ✅ Resolved in commit 40194d6.

**Changes:**
- Moved ACTIVITY_FILTER_MAX_SCAN_LIMIT and ACTIVITY_FILTER_CONCURRENCY_LIMIT to @sourcebot/shared/src/constants.ts
- Updated actions.ts to import constants from shared package
- Follows repository best practices for centralized configuration

File: packages/web/src/features/search/searchApi.ts [hunk 5]
```diff
@@ -127,8 +127,70 @@ const getFileWebUrl = (template: string, branch: string, fileName: string): stri
...
+                    files: [],
+                    repositoryInfo: [],
+                    isBranchFilteringEnabled: false,
+                    isSearchExhaustive: true,
+                } satisfies SearchResponse;
+            }
+
+            const repoFilter = timeFilterRepos.map(r => `repo:${r.name}`).join(" or ");
+            query = `(${query}) and (${repoFilter})`;
+        }
```
[the remark this answers is not in the document]
Comment (waynesun09) [reply packages/web/src/features/search/searchApi.ts]: ✅ Resolved - indexedAt filtering uses gte/lte (inclusive semantics).

**Rationale:**
- since/until parameters use inclusive semantics by convention
- Matches Git's --since and --until behavior
- Prevents edge case exclusions at exact timestamps
- Confirmed working with all 106 tests passing

[the remark this answers is not in the document]
Comment (waynesun09) [reply packages/web/src/features/search/searchApi.ts]: The includeDeletedFiles is removed

File: packages/mcp/src/index.ts [hunk 6]
```diff
@@ -5,7 +5,7 @@ import { McpServer } from '@modelcontextprotocol/sdk/server/mcp.js';
 import { StdioServerTransport } from '@modelcontextprotocol/sdk/server/stdio.js';
 import escapeStringRegexp from 'escape-string-regexp';
 import { z } from 'zod';
-import { listRepos, search, getFileSource } from './client.js';
+import { listRepos, search, getFileSource, searchCommits } from './client.js';
```
[the remark this answers is not in the document]
Comment (waynesun09) [reply packages/mcp/src/index.ts]: resolved in b30bbd3, dual support with repo id and name

@brendan-kellam the current list_repos return id which is repo name, don't want to change that field to be numeric id and add a new repo name field which might cause inconsistent issue. 

What do you think?

File: packages/web/src/features/search/schemas.ts (hunk 1 above)
[the remark this answers is not in the document]
Comment (waynesun09) [reply packages/web/src/features/search/schemas.ts]: resolved in 0071ca6

File: packages/shared/src/constants.ts [hunk 7]
```diff
@@ -30,3 +30,9 @@ export const DEFAULT_CONFIG_SETTINGS: ConfigSettings = {
     experiment_repoDrivenPermissionSyncIntervalMs: 1000 * 60 * 60 * 24, // 24 hours
     experiment_userDrivenPermissionSyncIntervalMs: 1000 * 60 * 60 * 24, // 24 hours
 }
+
+import { env } from "./env.server.js";
+import path from "path";
+
+export const REPOS_CACHE_DIR = path.join(env.DATA_CACHE_DIR, 'repos');
+export const INDEX_CACHE_DIR = path.join(env.DATA_CACHE_DIR, 'index');
```
[the remark this answers is not in the document]
Comment (waynesun09) [reply packages/shared/src/constants.ts]: resolved in 1bdf681

File: packages/web/src/app/api/(server)/commits/route.ts [hunk 8]
```diff
@@ -0,0 +1,27 @@
...
+
+    if (!parsed.success) {
+        return serviceErrorResponse(
+            schemaValidationError(parsed.error)
+        );
+    }
+
+    const result = await withOptionalAuthV2(async () => {
+        return await searchCommits(parsed.data as SearchCommitsRequest);
+    });
```
Comment (DaoDaoNoCode) [inline packages/web/src/app/api/(server)/commits/route.ts]: Nit:
1. The `as` assertion is unnecessary because the `safeParseAsync` guards the type.
2. The `async` and `await` in the callback functions are also unnecessary because the `searchCommits` is the only function you are calling here, so you are not doing any synchronous calls.
3. The `withOptionalAuthV2` is redundant here because you already have it in the `searchCommits` method. 

```suggestion
    const result = await searchCommits(parsed.data);
```

File: packages/web/src/actions.ts [hunk 9]
```diff
@@ -460,23 +467,86 @@ export const getUserApiKeys = async (domain: string): Promise<{ name: string; cr
...
+                    // Handle successful result
+                    if (Array.isArray(result)) {
+                        return result.length > 0 ? repo : null;
+                    }
+
+                    // Handle ServiceError
+                    const errorMessage = result.message ?? '';
+                    if (!errorMessage.includes('not found on Sourcebot server disk')) {
+                        // Log unexpected errors, but not "repo not cloned yet" errors
+                        console.error(
```
Comment (DaoDaoNoCode) [inline packages/web/src/actions.ts]: Maybe change it to `logger.error` to keep the consistency of the error logs with other places in the file since it's a server-side error?

File: packages/web/src/app/api/(server)/commits/route.ts (hunk 8 above)
Comment (waynesun09) [reply packages/web/src/app/api/(server)/commits/route.ts]: Done, thanks

File: packages/web/src/actions.ts (hunk 9 above)
Comment (waynesun09) [reply packages/web/src/actions.ts]: Done

**Comment by waynesun09** [comment]
@coderabbitai please review again

**Comment by waynesun09** [comment]
@coderabbitai please review again

**Comment by brendan-kellam** [comment]
Hey @waynesun09 sorry for the delay - I will get to this tomorrow

**Comment by waynesun09** [comment]
@brendan-kellam no worries, I've rebased to latest main branch

**Comment by waynesun09** [comment]
@brendan-kellam any free time to check, thanks?

File: packages/web/src/features/search/types.ts [hunk 10]
```diff
@@ -104,6 +107,7 @@ export const searchResponseSchema = z.object({
     files: z.array(searchFileSchema),
     repositoryInfo: z.array(repositoryInfoSchema),
     isSearchExhaustive: z.boolean(),
+    isBranchFilteringEnabled: z.boolean().optional(), // Whether branch filtering is enabled for this search.
```
Comment (brendan-kellam) [inline packages/web/src/features/search/types.ts]: I think this might be a artifact of a merge conflict - we don't actually need this in the response body anymore I don't think.

File: packages/web/src/features/search/searchApi.ts (hunk 2 above)
Comment (brendan-kellam) [reply packages/web/src/features/search/searchApi.ts]: Apologies for the very late reply. IMO I don't think this is a valuable enough use case - I'm not sure in what scenario you would care when a repository was added to Sourcebot. I think the more relevant thing to know is "what code was just updated?" (i.e., last committed to).

We could store the commit time of the default branch in the db and use that, but I'll leave that as a separate PR.

File: packages/web/src/features/search/types.ts [hunk 11]
```diff
@@ -89,6 +89,9 @@ export const searchOptionsSchema = z.object({
     whole: z.boolean().optional(),                    // Whether to return the whole file as part of the response.
     isRegexEnabled: z.boolean().optional(),           // Whether to enable regular expression search.
     isCaseSensitivityEnabled: z.boolean().optional(), // Whether to enable case sensitivity.
+    gitRevision: z.string().optional(),               // Filter by git branch/revision.
```
Comment (brendan-kellam) [inline packages/web/src/features/search/types.ts]: instead of passing the revision as a request param, the MCP server can just append the `rev:` prefix to the query.

File: packages/mcp/src/index.ts (hunk 6 above)
[the remark this answers is not in the document]
Comment (brendan-kellam) [reply packages/mcp/src/index.ts]: I think we should just go with the repo name instead of the repo id.
